src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc

```diff
@@ -16,41 +16,261 @@
+                    interested_parties,
+                    grpc_slice_from_static_string(kS2AServiceMethod), &slice,
+                    GRPC_MILLIS_INF_FUTURE, /*reserved=*/nullptr);
+  GRPC_CLOSURE_INIT(&on_handshaker_service_resp_recv_, grpc_cb, this,
+                    grpc_schedule_on_exec_ctx);
+  GRPC_CLOSURE_INIT(&on_status_received_, s2a_on_status_received, this,
+                    grpc_schedule_on_exec_ctx);
+  grpc_slice_unref_internal(slice);
+}
+
 tsi_result s2a_handshaker_client_create(
     s2a_tsi_handshaker* handshaker, grpc_channel* channel,
-    const char* handshaker_service_url, grpc_pollset_set* interested_parties,
-    grpc_s2a_credentials_options* options, const grpc_slice& target_name,
-    grpc_iomgr_cb_func grpc_cb, tsi_handshaker_on_next_done_cb cb,
-    void* user_data, bool is_client, s2a_handshaker_client** client,
-    char** error_details) {
-  return TSI_UNIMPLEMENTED;
+    grpc_pollset_set* interested_parties, grpc_s2a_credentials_options* options,
+    const grpc_slice& target_name, grpc_iomgr_cb_func grpc_cb,
+    tsi_handshaker_on_next_done_cb cb, void* user_data, bool is_client,
+    s2a_handshaker_client** client) {
+  if (channel == nullptr || client == nullptr || options == nullptr ||
+      options->handshaker_service_url().empty()) {
+    gpr_log(GPR_ERROR, kS2AHandshakerClientNullptrArguments);
+    return TSI_INVALID_ARGUMENT;
+  }
+  *client = new s2a_handshaker_client(handshaker, channel, interested_parties,
+                                      options, target_name, grpc_cb, cb,
+                                      user_data, is_client);
+  return TSI_OK;
+}
+
+void s2a_handshaker_client::shutdown() {
+  if (call_ != nullptr) {
+    grpc_call_cancel_internal(call_);
+  }
+}
+
+/** This method destroys the grpc_call owned by the s2a_handshaker_client. **/
+static void s2a_handshaker_call_unref(void* arg, grpc_error* error) {
+  grpc_call* call = static_cast<grpc_call*>(arg);
+  grpc_call_unref(call);
 }
 
-void s2a_handshaker_client_destroy(s2a_handshaker_client* client) { return; }
+s2a_handshaker_client::~s2a_handshaker_client() {
+  if (gpr_unref(&refs_)) {
+    if (call_ != nullptr) {
+      grpc_core::ExecCtx::Run(
+          DEBUG_LOCATION,
+          GRPC_CLOSURE_CREATE(s2a_handshaker_call_unref, call_,
+                              grpc_schedule_on_exec_ctx),
+          GRPC_ERROR_NONE);
+    }
+    grpc_byte_buffer_destroy(send_buffer_);
+    grpc_byte_buffer_destroy(recv_buffer_);
+    send_buffer_ = nullptr;
+    recv_buffer_ = nullptr;
+    grpc_metadata_array_destroy(&recv_initial_metadata_);
+    grpc_slice_unref_internal(recv_bytes_);
+    grpc_slice_unref_internal(target_name_);
+    grpc_s2a_credentials_options_destroy(options_);
+    gpr_free(buffer_);
+    grpc_slice_unref_internal(handshake_status_details_);
+    gpr_mu_destroy(&mu_);
+  }
+}
+
+void s2a_handshaker_client_destroy(s2a_handshaker_client* client) {
```

Comment:
Same comments as `s2a_handshaker_client_create()`. It seems redundant now that we have a destructor.

Reply:
Same as above.